lib/demuxeres/gstdemuxeres.c

```diff
@@ -0,0 +1,721 @@
+
+struct _GstDemuxerESPrivate
+{
+  GstElement *pipeline;
+  GstElement *funnel;
+  GstElement *appsink;
+
+  gint current_stream_id;
+  gint current_stream_type;
+
+  GList *streams;
+
+  GstDemuxerESState state;
+  GCond ready_cond;
+  GMutex ready_mutex;
+
+  GThread *bus_thread;
+  gboolean bus_exit;
+  GstSample *pending_sample;
+};
+
+
+static gchar *
+get_gst_valid_uri (const gchar * filename)
+{
+  if (gst_uri_is_valid (filename)) {
+    return g_strdup (filename);
+  }
+
+  return gst_filename_to_uri (filename, NULL);
+}
+
+struct _GstDemuxerESPacketPrivate
+{
+  GstSample *sample;
+};
+
+
+static inline void
+set_demuxer_state (GstDemuxerES * demuxer, GstDemuxerESState state)
+{
+  GstDemuxerESPrivate *priv = demuxer->priv;
+  g_mutex_lock (&priv->ready_mutex);
+  GST_LOG ("Set state %d", state);
+  priv->state = state;
+  g_cond_signal (&priv->ready_cond);
+  g_mutex_unlock (&priv->ready_mutex);
+}
+
+
+static GstDemuxerEStream *
+find_stream (GstDemuxerES * demuxer, const gchar * stream_id)
+{
+  GstDemuxerESPrivate *priv = demuxer->priv;
+  GList *l;
+
+  for (l = priv->streams; l != NULL; l = g_list_next (l)) {
+    GstDemuxerEStream *stream = l->data;
+    if (!g_strcmp0 (stream->stream_id, stream_id)) {
+      return stream;
+    }
+  }
+  return NULL;
+}
+
+static gboolean
+appsink_handle_event (GstDemuxerES * demuxer, GstEvent * event)
+{
+  GstDemuxerESPrivate *priv = demuxer->priv;
+  gboolean ret = FALSE;
+  GST_LOG ("Event %s", GST_EVENT_TYPE_NAME (event));
+  switch (GST_EVENT_TYPE (event)) {
+    case GST_EVENT_STREAM_START:{
+      const gchar *stream_id;
+      GstDemuxerEStream *stream;
+      gst_event_parse_stream_start (event, &stream_id);
+      stream = find_stream (demuxer, stream_id);
+      g_assert (stream);
+      priv->current_stream_id = stream->id;
+      priv->current_stream_type = stream->type;
+      ret = TRUE;
+      break;
+    }
+    case GST_EVENT_STREAM_GROUP_DONE:{
+      set_demuxer_state (demuxer, DEMUXER_ES_STATE_EOS);
+      ret = TRUE;
+      break;
+    }
+    default:
+      GST_LOG ("Unhandled event %s", GST_EVENT_TYPE_NAME (event));
+  }
+  gst_event_unref (event);
+  return ret;
+}
+
+static GstDemuxerESPacket *
+appsink_read_packet (GstDemuxerES * demuxer)
+{
+  GstSample *sample = NULL;
+  GstBuffer *buffer;
+  GstDemuxerESPacket *packet = NULL;
+  GstDemuxerESPrivate *priv = demuxer->priv;
+  GstMiniObject *object;
+
+  if (!priv->pending_sample) {
+    while ((object =
+            gst_app_sink_try_pull_object (GST_APP_SINK (priv->appsink), -1))) {
+      if (GST_IS_EVENT (object)) {
+        appsink_handle_event (demuxer, GST_EVENT (object));
+      } else if (GST_IS_SAMPLE (object)) {
+        sample = GST_SAMPLE (object);
+        break;
+      }
+    }
+  } else {
+    sample = priv->pending_sample;
+    priv->pending_sample = NULL;
+  }
+
+  if (sample) {
+    buffer = gst_sample_get_buffer (sample);
+    if (buffer) {
+      packet = g_new0 (GstDemuxerESPacket, 1);
```

Comment:
If you're sure that the packet is going to be fulled defined, there's no need to do g_new0, since it zeroes the memory, but later you will write it all over again. I was bitten that this issue in the past, is a huge CPU swallower when you are allocating a lot of packets constantly. Better just g_new and be sure to fill the whole structure with:

```
*packet = (GstDemuxerESPacket) {<
...
};
```